File: app/lib/services/formula.dart
/// A small, closed arithmetic language for admin-defined calculators.
///
/// This is the Dart twin of `api/lib/utils/formula.ts`. The two implement the
/// same grammar and must stay in step: the server validates a formula when an
/// admin writes it, and this evaluates the same text on a phone, possibly
/// months later and with no connection. A formula the server accepted must
/// never mean something different here.
///
/// The app evaluates rather than asking the server to, because working out a
/// mealtime dose in a kitchen with no signal is the normal case for these
/// families, not an edge case.
///
/// ## The language
///
///   number        12, 3.5, .5 is rejected as malformed
///   variable      tdd, carbs, glucose
///   operators     + - * / ( ) and unary minus
///   functions     min(a, b), max(a, b), round(x), floor(x), ceil(x)
///
/// Nothing else parses. There is no string, no comparison, no assignment and
/// no property access, so a stored formula cannot express anything but
/// arithmetic.
library;

/// Raised when a formula cannot be parsed or cannot be worked out. The
/// message is written for a parent to read, not a developer.
class FormulaException implements Exception {
  final String message;

  FormulaException(this.message);

  @override
  String toString() => message;
}

enum _TokenType { number, identifier, operator, paren, comma }

class _Token {
  final _TokenType type;
  final String value;

  _Token(this.type, this.value);
}

/// Functions a formula may call, with their exact arity.
final Map<String, ({int arity, double Function(List<double>) apply})>
_functions = {
  'min': (arity: 2, apply: (args) => args[0] < args[1] ? args[0] : args[1]),
  'max': (arity: 2, apply: (args) => args[0] > args[1] ? args[0] : args[1]),
  'round': (arity: 1, apply: (args) => args[0].roundToDouble()),
  'floor': (arity: 1, apply: (args) => args[0].floorToDouble()),
  'ceil': (arity: 1, apply: (args) => args[0].ceilToDouble()),
};

const int _maxLength = 500;

final _whitespace = RegExp(r'\s');
final _digitOrDot = RegExp(r'[0-9.]');
final _identifierStart = RegExp(r'[A-Za-z_]');
final _identifierPart = RegExp(r'[A-Za-z0-9_]');

List<_Token> _tokenize(String source) {
  final tokens = <_Token>[];
  var i = 0;

  while (i < source.length) {
    final char = source[i];

    if (_whitespace.hasMatch(char)) {
      i++;
      continue;
    }

    if (_digitOrDot.hasMatch(char)) {
      final start = i;
      while (i < source.length && _digitOrDot.hasMatch(source[i])) {
        i++;
      }
      final value = source.substring(start, i);
      if ('.'.allMatches(value).length > 1 || value == '.') {
        throw FormulaException('"$value" is not a number.');
      }
      tokens.add(_Token(_TokenType.number, value));
      continue;
    }

    if (_identifierStart.hasMatch(char)) {
      final start = i;
      while (i < source.length && _identifierPart.hasMatch(source[i])) {
        i++;
      }
      tokens.add(_Token(_TokenType.identifier, source.substring(start, i)));
      continue;
    }

    if ('+-*/'.contains(char)) {
      tokens.add(_Token(_TokenType.operator, char));
      i++;
      continue;
    }

    if (char == '(' || char == ')') {
      tokens.add(_Token(_TokenType.paren, char));
      i++;
      continue;
    }

    if (char == ',') {
      tokens.add(_Token(_TokenType.comma, char));
      i++;
      continue;
    }

    throw FormulaException('"$char" cannot be used in a formula.');
  }

  return tokens;
}

/// Recursive-descent parser, matching the precedence in formula.ts:
///   expression := term (("+" | "-") term)*
///   term       := factor (("*" | "/") factor)*
///   factor     := ("-")? primary
///   primary    := number | identifier | call | "(" expression ")"
class _Parser {
  final List<_Token> _tokens;
  final Map<String, double> _variables;
  int _index = 0;

  _Parser(this._tokens, this._variables);

  _Token? _peek() => _index < _tokens.length ? _tokens[_index] : null;

  _Token? _next() => _index < _tokens.length ? _tokens[_index++] : null;

  double parse() {
    final value = _expression();
    final leftover = _peek();
    if (leftover != null) {
      throw FormulaException('Unexpected "${leftover.value}" in the formula.');
    }
    return value;
  }

  double _expression() {
    var value = _term();
    while (true) {
      final token = _peek();
      if (token == null ||
          token.type != _TokenType.operator ||
          (token.value != '+' && token.value != '-')) {
        break;
      }
      _next();
      final right = _term();
      value = token.value == '+' ? value + right : value - right;
    }
    return value;
  }

  double _term() {
    var value = _factor();
    while (true) {
      final token = _peek();
      if (token == null ||
          token.type != _TokenType.operator ||
          (token.value != '*' && token.value != '/')) {
        break;
      }
      _next();
      final right = _factor();
      if (token.value == '*') {
        value *= right;
      } else {
        // Matches formula.ts: a division by zero is refused outright rather
        // than producing infinity, so a parent never sees a nonsense dose.
        if (right == 0) {
          throw FormulaException('This works out as a division by zero.');
        }
        value /= right;
      }
    }
    return value;
  }

  double _factor() {
    final token = _peek();
    if (token != null &&
        token.type == _TokenType.operator &&
        token.value == '-') {
      _next();
      return -_factor();
    }
    if (token != null &&
        token.type == _TokenType.operator &&
        token.value == '+') {
      _next();
      return _factor();
    }
    return _primary();
  }

  double _primary() {
    final token = _next();
    if (token == null) throw FormulaException('The formula ends too early.');

    if (token.type == _TokenType.number) {
      final parsed = double.tryParse(token.value);
      if (parsed == null) {
        throw FormulaException('"${token.value}" is not a number.');
      }
      return parsed;
    }

    if (token.type == _TokenType.identifier) {
      final fn = _functions[token.value];
      if (fn != null) return _call(token.value, fn);

      final value = _variables[token.value];
      if (value == null) {
        throw FormulaException('No value was given for "${token.value}".');
      }
      if (!value.isFinite) {
        throw FormulaException(
          'The value for "${token.value}" is not a number.',
        );
      }
      return value;
    }

    if (token.type == _TokenType.paren && token.value == '(') {
      final value = _expression();
      final closing = _next();
      if (closing?.value != ')') {
        throw FormulaException('A bracket is not closed.');
      }
      return value;
    }

    throw FormulaException('Unexpected "${token.value}" in the formula.');
  }

  double _call(
    String name,
    ({int arity, double Function(List<double>) apply}) fn,
  ) {
    final open = _next();
    if (open?.value != '(') {
      throw FormulaException('"$name" must be followed by a bracket.');
    }

    final args = <double>[];
    if (_peek()?.value == ')') {
      _next();
    } else {
      while (true) {
        args.add(_expression());
        final separator = _next();
        if (separator?.value == ')') break;
        if (separator?.value != ',') {
          throw FormulaException('"$name" is missing a bracket.');
        }
      }
    }

    if (args.length != fn.arity) {
      throw FormulaException(
        '"$name" needs ${fn.arity} number${fn.arity == 1 ? '' : 's'}, not ${args.length}.',
      );
    }
    return fn.apply(args);
  }
}

/// Works out [expression] using [variables].
///
/// Throws [FormulaException] with a message suitable for showing directly.
double evaluateFormula(String expression, Map<String, double> variables) {
  if (expression.length > _maxLength) {
    throw FormulaException(
      'A formula cannot be longer than $_maxLength characters.',
    );
  }

  final result = _Parser(_tokenize(expression), variables).parse();

  if (!result.isFinite) {
    throw FormulaException(
      'This formula does not work out to a usable number.',
    );
  }
  return result;
}

File: api/lib/utils/formula.ts
/**
 * A small, closed arithmetic language for admin-defined calculators.
 *
 * ## Why this exists rather than `eval`
 *
 * Calculators in this product turn parent-entered numbers into insulin
 * guidance. The formulas are written by study staff through the dashboard,
 * stored in the database, and evaluated on a child's phone. That makes a
 * stored formula an input crossing a trust boundary twice over, so it is
 * parsed by an explicit grammar that can express arithmetic and nothing else.
 * `eval`, `new Function`, and every "tiny expression library" that compiles to
 * them are out of the question: a formula field that can reach `fetch` or
 * `localStorage` is a stored-XSS hole wearing a lab coat.
 *
 * ## The language
 *
 *   number        12, 3.5, .5
 *   variable      tdd, carbs, glucose  (letters, digits, underscore; not
 *                                       leading-digit)
 *   operators     + - * / ( ) and unary minus
 *   functions     min(a, b), max(a, b), round(x), floor(x), ceil(x)
 *
 * No assignment, no comparison, no strings, no property access, no calls to
 * anything not in the table above. An unknown identifier is a parse error, not
 * an implicit `undefined` — a typo in a dosing formula must fail loudly at the
 * moment it is written, not produce NaN on a phone months later.
 *
 * ## Dart counterpart
 *
 * app/lib/services/formula.dart implements the same grammar so the app can
 * evaluate offline. The two must stay in step; both are covered by tests
 * asserting identical results for the same expressions.
 */

export class FormulaError extends Error {}

type TokenType = "number" | "identifier" | "operator" | "paren" | "comma";

interface Token {
  type: TokenType;
  value: string;
  position: number;
}

/** Functions a formula may call, with their exact arity. */
const FUNCTIONS: Record<string, { arity: number; apply: (args: number[]) => number }> = {
  min: { arity: 2, apply: ([a, b]) => Math.min(a, b) },
  max: { arity: 2, apply: ([a, b]) => Math.max(a, b) },
  round: { arity: 1, apply: ([a]) => Math.round(a) },
  floor: { arity: 1, apply: ([a]) => Math.floor(a) },
  ceil: { arity: 1, apply: ([a]) => Math.ceil(a) },
};

export const FORMULA_FUNCTION_NAMES = Object.keys(FUNCTIONS);

/** Longest expression accepted, as a guard against pathological input. */
const MAX_LENGTH = 500;

function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;

  while (i < source.length) {
    const char = source[i];

    if (/\s/.test(char)) {
      i += 1;
      continue;
    }

    if (/[0-9.]/.test(char)) {
      const start = i;
      while (i < source.length && /[0-9.]/.test(source[i])) i += 1;
      const value = source.slice(start, i);
      // "1.2.3" tokenizes as one run of digits-and-dots; reject it here
      // rather than letting Number() quietly return NaN.
      if ((value.match(/\./g) ?? []).length > 1 || value === ".") {
        throw new FormulaError(`"${value}" is not a number.`);
      }
      tokens.push({ type: "number", value, position: start });
      continue;
    }

    if (/[A-Za-z_]/.test(char)) {
      const start = i;
      while (i < source.length && /[A-Za-z0-9_]/.test(source[i])) i += 1;
      tokens.push({ type: "identifier", value: source.slice(start, i), position: start });
      continue;
    }

    if ("+-*/".includes(char)) {
      tokens.push({ type: "operator", value: char, position: i });
      i += 1;
      continue;
    }

    if (char === "(" || char === ")") {
      tokens.push({ type: "paren", value: char, position: i });
      i += 1;
      continue;
    }

    if (char === ",") {
      tokens.push({ type: "comma", value: char, position: i });
      i += 1;
      continue;
    }

    throw new FormulaError(`"${char}" cannot be used in a formula.`);
  }

  return tokens;
}

/**
 * Recursive-descent parser producing a value directly.
 *
 * Grammar, lowest precedence first:
 *   expression := term (("+" | "-") term)*
 *   term       := factor (("*" | "/") factor)*
 *   factor     := ("-")? primary
 *   primary    := number | identifier | call | "(" expression ")"
 */
class Parser {
  private index = 0;

  constructor(
    private readonly tokens: Token[],
    private readonly variables: Readonly<Record<string, number>>,
    /** When set, identifiers are checked but arithmetic is not performed. */
    private readonly declaredOnly: ReadonlySet<string> | null,
  ) {}

  private peek(): Token | undefined {
    return this.tokens[this.index];
  }

  private next(): Token | undefined {
    return this.tokens[this.index++];
  }

  parse(): number {
    const value = this.expression();
    const leftover = this.peek();
    if (leftover) {
      throw new FormulaError(`Unexpected "${leftover.value}" in the formula.`);
    }
    return value;
  }

  private expression(): number {
    let value = this.term();
    for (;;) {
      const token = this.peek();
      if (token?.type !== "operator" || (token.value !== "+" && token.value !== "-")) break;
      this.next();
      const right = this.term();
      value = token.value === "+" ? value + right : value - right;
    }
    return value;
  }

  private term(): number {
    let value = this.factor();
    for (;;) {
      const token = this.peek();
      if (token?.type !== "operator" || (token.value !== "*" && token.value !== "/")) break;
      this.next();
      const right = this.factor();
      if (token.value === "*") {
        value *= right;
      } else {
        // Dividing by zero is the single most likely way an admin-authored
        // dosing formula goes wrong (a total daily dose of 0 is a perfectly
        // ordinary thing for a parent to type). Refusing here means the app
        // shows "check your numbers", never "Infinity units".
        if (right === 0) throw new FormulaError("This works out as a division by zero.");
        value /= right;
      }
    }
    return value;
  }

  private factor(): number {
    const token = this.peek();
    if (token?.type === "operator" && token.value === "-") {
      this.next();
      return -this.factor();
    }
    if (token?.type === "operator" && token.value === "+") {
      this.next();
      return this.factor();
    }
    return this.primary();
  }

  private primary(): number {
    const token = this.next();
    if (!token) throw new FormulaError("The formula ends too early.");

    if (token.type === "number") return Number(token.value);

    if (token.type === "identifier") {
      const fn = FUNCTIONS[token.value];
      if (fn) return this.call(token.value, fn);

      if (this.declaredOnly) {
        if (!this.declaredOnly.has(token.value)) {
          throw new FormulaError(`"${token.value}" is not one of this calculator's inputs.`);
        }
        // Checking only: a stand-in that exercises every branch without
        // pretending to be a real measurement.
        return 1;
      }

      const value = this.variables[token.value];
      if (value === undefined) {
        throw new FormulaError(`No value was given for "${token.value}".`);
      }
      if (!Number.isFinite(value)) {
        throw new FormulaError(`The value for "${token.value}" is not a number.`);
      }
      return value;
    }

    if (token.type === "paren" && token.value === "(") {
      const value = this.expression();
      const closing = this.next();
      if (closing?.value !== ")") throw new FormulaError("A bracket is not closed.");
      return value;
    }

    throw new FormulaError(`Unexpected "${token.value}" in the formula.`);
  }

  private call(name: string, fn: (typeof FUNCTIONS)[string]): number {
    const open = this.next();
    if (open?.value !== "(") throw new FormulaError(`"${name}" must be followed by a bracket.`);

    const args: number[] = [];
    if (this.peek()?.value === ")") {
      this.next();
    } else {
      for (;;) {
        args.push(this.expression());
        const separator = this.next();
        if (separator?.value === ")") break;
        if (separator?.value !== ",") throw new FormulaError(`"${name}" is missing a bracket.`);
      }
    }

    if (args.length !== fn.arity) {
      throw new FormulaError(
        `"${name}" needs ${fn.arity} number${fn.arity === 1 ? "" : "s"}, not ${args.length}.`,
      );
    }
    return fn.apply(args);
  }
}

/**
 * Computes a formula against a set of named values.
 *
 * Throws {@link FormulaError} with a message written for the person who typed
 * the formula, not for a developer — these surface directly in the dashboard.
 */
export function evaluateFormula(
  expression: string,
  variables: Readonly<Record<string, number>>,
): number {
  if (expression.length > MAX_LENGTH) {
    throw new FormulaError(`A formula cannot be longer than ${MAX_LENGTH} characters.`);
  }

  const result = new Parser(tokenize(expression), variables, null).parse();

  if (!Number.isFinite(result)) {
    throw new FormulaError("This formula does not work out to a usable number.");
  }
  return result;
}

/**
 * Checks a formula at authoring time.
 *
 * Confirms it parses and that every name in it is one of the calculator's own
 * inputs (or an earlier output). Returns the error message to show, or null
 * when the formula is sound.
 */
export function validateFormula(expression: string, availableNames: readonly string[]): string | null {
  if (!expression.trim()) return "Enter a formula.";
  if (expression.length > MAX_LENGTH) {
    return `A formula cannot be longer than ${MAX_LENGTH} characters.`;
  }

  try {
    new Parser(tokenize(expression), {}, new Set(availableNames)).parse();
    return null;
  } catch (error) {
    if (error instanceof FormulaError) {
      // A division by zero found while checking is an artefact of the
      // stand-in values, not a fault in the formula — a real division by
      // zero is caught when it actually happens.
      if (error.message.includes("division by zero")) return null;
      return error.message;
    }
    throw error;
  }
}
